src/allotropy/allotrope/schema_parser/generate_schemas.py

```diff
@@ -84,45 +85,56 @@ def _generate_schema(model_path: Path, schema_path: Path) -> None:
     lint_file(str(model_path))
 
 
-def _should_generate_schema(schema_path: str, schema_regex: str | None = None) -> bool:
+def _should_generate_schema(schema_path: Path, schema_regex: str | None = None) -> bool:
     # Skip files in the shared directory
-    if schema_path.startswith("shared"):
+    rel_schema_path = str(get_rel_schema_path(schema_path))
+    if rel_schema_path.startswith("shared"):
         return False
-    if is_backup_file(schema_path):
+    if is_backup_file(rel_schema_path):
         return False
     if schema_regex:
-        return bool(re.match(schema_regex, str(schema_path)))
+        return bool(re.match(schema_regex, str(rel_schema_path)))
     return True
 
 
+def make_model_directories(model_path: Path) -> None:
+    if model_path == MODEL_DIR_PATH:
+        return
+
+    make_model_directories(model_path.parent)
+    if model_path.exists():
+        return
+    model_path.mkdir()
+    init_path = Path(model_path, "__init__.py")
+    if not init_path.exists():
+        init_path.touch()
+
+
 def generate_schemas(
-    root_dir: Path,
     *,
     dry_run: bool | None = False,
     schema_regex: str | None = None,
 ) -> list[str]:
     """Generate schemas from JSON schema files.
-
-    :root_dir: The root directory of the project.
     :dry_run: If true, does not save changes to any models, but still returns the list of models that would change.
     :schema_regex: If set, filters schemas to generate using regex.
     :return: A list of model files that were changed.
     """
-
     unit_to_iri: dict[str, str] = {}
     with backup(GENERATED_SHARED_PATHS, restore=dry_run):
-        os.chdir(os.path.join(root_dir, SCHEMA_DIR_PATH))
```

Comment:
Note to reviewer: I realized that with SCHEMA_DIR_PATH being an absolute path from path_util, there is no need for root_dir/chdir